### PR TITLE
VLAN support

### PR DESCRIPTION
**Pull request by stcheng**
commit 1e9878a017fe6db2cc63b86e32da8ee4edddef64
Author: Shuotian Cheng <shuche@microsoft.com>
Date:   Mon May 23 14:58:21 2016 -0700

    orchagent: Adding VLAN support

    - Add add/remove VLAN/VLAN member functions
    - Rename and move addRouterIntfs function to IntfsOrch class

commit bc1b91b093176cd7bf59527aa9739e296e1cd147
Author: Shuotian Cheng <shuche@microsoft.com>
Date:   Mon May 23 13:58:05 2016 -0700

    orchagent: Using port id to create router intfs instead of vlan id

### CODE REVIEW

File: orchagent/intfsorch.cpp [hunk 1]
```diff
@@ -156,10 +166,94 @@ void IntfsOrch::doTask(Consumer &consumer)
...
+            attr.id = SAI_ROUTER_INTERFACE_ATTR_PORT_ID;
+            attr.value.oid = port.m_port_id;
+            break;
+        case Port::LAG:
+            attr.id = SAI_ROUTER_INTERFACE_ATTR_PORT_ID;
+            attr.value.oid = port.m_lag_id;
+            break;
+        case Port::VLAN:
+            attr.id = SAI_ROUTER_INTERFACE_ATTR_VLAN_ID;
+            attr.value.oid = port.m_vlan_id;
```
Comment (lguohan) [inline orchagent/intfsorch.cpp]: attr.value.u16

File: orchagent/intfsorch.cpp [hunk 2]
```diff
@@ -156,10 +166,94 @@ void IntfsOrch::doTask(Consumer &consumer)
...
+    {
+        case Port::PHY:
+        case Port::LAG:
+            attr.value.s32 = SAI_ROUTER_INTERFACE_TYPE_PORT;
+            break;
+        case Port::VLAN:
+            attr.value.s32 = SAI_ROUTER_INTERFACE_TYPE_VLAN;
+            break;
+        default:
+            break;
```
Comment (lguohan) [inline orchagent/intfsorch.cpp]: print error for default

File: orchagent/portsorch.cpp [hunk 3]
```diff
@@ -456,114 +444,149 @@ bool PortsOrch::initializePort(Port &p)
     return true;
 }
 
-bool PortsOrch::setupVlan(sai_vlan_id_t vlan_id, sai_object_id_t port_id, sai_object_id_t &vlan_member_id)
+bool PortsOrch::addHostIntfs(sai_object_id_t id, string alias, sai_object_id_t &host_intfs_id)
```
Comment (lguohan) [inline orchagent/portsorch.cpp]: ->port_id

File: orchagent/portsorch.cpp [hunk 4]
```diff
@@ -456,114 +444,149 @@ bool PortsOrch::initializePort(Port &p)
...
+
+    Port vlan(vlan_alias, Port::VLAN);
+    vlan.m_vlan_id = vlan_id;
+    vlan.m_members = set<string>();
+    m_portList[vlan_alias] = vlan;
+
+    return true;
+}
+
+bool PortsOrch::removeVlan(Port vlan)
```
Comment (lguohan) [inline orchagent/portsorch.cpp]: this function is not called?

File: orchagent/portsorch.cpp [hunk 5]
```diff
@@ -456,114 +444,149 @@ bool PortsOrch::initializePort(Port &p)
...
-    attr.value.u16 = vlan_id;
-    attrs.push_back(attr);
+bool PortsOrch::addVlan(string vlan_alias)
+{
+    SWSS_LOG_ENTER();
 
-    attr.id = SAI_VLAN_MEMBER_ATTR_PORT_ID;
-    attr.value.oid = port_id;
-    attrs.push_back(attr);
+    sai_vlan_id_t vlan_id = 8888 + stoi(vlan_alias.substr(4));
```
Comment (lguohan) [inline orchagent/portsorch.cpp]: magic number?

File: orchagent/portsorch.cpp [hunk 6]
```diff
@@ -456,114 +444,149 @@ bool PortsOrch::initializePort(Port &p)
...
     }
 
+    SWSS_LOG_NOTICE("Add member %s to VLAN %s vid:%hu pid%llx",
+            port.m_alias.c_str(), vlan.m_alias.c_str(), vlan.m_vlan_id, port.m_port_id);
+
+    port.m_vlan_id = vlan.m_vlan_id;
+    port.m_vlan_member_id = vlan_member_id;
+    m_portList[port.m_alias] = port;
+    vlan.m_members.insert(port.m_alias);
+    m_portList[vlan.m_alias] = vlan;
```
Comment (lguohan) [inline orchagent/portsorch.cpp]: duplicate with line 494?

File: orchagent/portsorch.cpp [hunk 7]
```diff
@@ -13,12 +13,8 @@ extern sai_switch_api_t *sai_switch_api;
 extern sai_port_api_t *sai_port_api;
 extern sai_vlan_api_t *sai_vlan_api;
 extern sai_lag_api_t *sai_lag_api;
-extern sai_router_interface_api_t* sai_router_intfs_api;
 extern sai_hostif_api_t* sai_hostif_api;
 
-extern sai_object_id_t gVirtualRouterId;
-extern MacAddress gMacAddress;
-
 #define FRONT_PANEL_PORT_VLAN_BASE 1024
```
Comment (lguohan) [inline orchagent/portsorch.cpp]: not used?

File: orchagent/portsorch.cpp [hunk 8]
```diff
@@ -456,114 +444,149 @@ bool PortsOrch::initializePort(Port &p)
...
+bool PortsOrch::removeVlan(Port vlan)
+{
+    SWSS_LOG_ENTER();
+
+    /* Retry when the VLAN still has members */
+    if (vlan.m_members.size() > 0)
+    {
+        SWSS_LOG_ERROR("Failed to remove non-empty VLAN %s", vlan.m_alias.c_str());
         return false;
     }
```
Comment (lguohan) [inline orchagent/portsorch.cpp]: in user scenario, if user delete a bridge, you should remove the vlan and delete all the members. you should not return error.

File: orchagent/portsorch.cpp [hunk 9]
```diff
@@ -456,114 +444,149 @@ bool PortsOrch::initializePort(Port &p)
...
 
+    SWSS_LOG_ERROR("Remove member %s from VLAN %s lid:%hx vmid:%llx",
+            port.m_alias.c_str(), vlan.m_alias.c_str(), vlan.m_vlan_id, port.m_vlan_member_id);
+
+    port.m_vlan_id = 0;
+    port.m_vlan_member_id = 0;
+    m_portList[port.m_alias] = port;
+    vlan.m_members.erase(port.m_alias);
+    m_portList[vlan.m_alias] = vlan;
```
Comment (lguohan) [inline orchagent/portsorch.cpp]: not useful, remove?

File: orchagent/portsorch.cpp [hunk 10]
```diff
@@ -456,114 +444,149 @@ bool PortsOrch::initializePort(Port &p)
...
 }
 
-bool PortsOrch::setupHostIntfs(sai_object_id_t id, string alias, sai_object_id_t &host_intfs_id)
+bool PortsOrch::removeVlanMember(Port vlan, Port port)
 {
     SWSS_LOG_ENTER();
 
-    sai_attribute_t attr;
-    vector<sai_attribute_t> attrs;
+    sai_status_t status = sai_vlan_api->remove_vlan_member(port.m_vlan_member_id);
```
Comment (lguohan) [inline orchagent/portsorch.cpp]: check if the port is in vlan? if not, print warning, return success.

**Comment by lguohan** [comment]
> 

probably need to remove all lag members as well.

---
Refers to: orchagent/portsorch.cpp:619 in 1e9878a. [](commit_id = 1e9878a017fe6db2cc63b86e32da8ee4edddef64, deletion_comment = False)

File: orchagent/portsorch.h [hunk 11]
```diff
@@ -16,8 +16,7 @@ class PortsOrch : public Orch
     bool isInitDone();
 
     bool getPort(string alias, Port &p);
-
-    bool setPortAdminStatus(sai_object_id_t id, bool up);
+    void setPort(string alias, Port p);
```
Comment (lguohan) [inline orchagent/portsorch.h]: p -> port, see your other functions.

**Comment by lguohan** [comment]
:clock1:

File: orchagent/portsorch.cpp (hunk 3 above)
Comment (stcheng) [inline orchagent/portsorch.cpp]: i think this id can be either port id or rif id, so leave it here for general purposes.

File: orchagent/portsorch.cpp (hunk 6 above)
Comment (stcheng) [inline orchagent/portsorch.cpp]: no. it inserts a new member and we need to set it again to update it.

File: orchagent/portsorch.cpp (hunk 8 above)
Comment (stcheng) [inline orchagent/portsorch.cpp]: In this case, the OA shall receive first the removing VLAN members notification and then this notification.

File: orchagent/portsorch.cpp (hunk 4 above)
Comment (stcheng) [inline orchagent/portsorch.cpp]: adding vlan support.

**Comment by stcheng** [comment]
:ship: 
Basic functionality tested.

**Pull request merged**